ServantHDBC/src/Api/Product.hs

```diff
@@ -12,80 +12,113 @@ module Api.Product (ProductAPI, productHandlers) where
 
 import Control.Monad.Trans.Except
 import Data.Aeson
+import Data.Text
+import Data.Time.Clock
 import GHC.Generics
 import Servant
+import Web.HttpApiData
+
 
 
 data Product = Product
-    { ids               :: String
-    , q                 :: String
-    , title             :: String
-    , sku               :: String
-    , prod_type         :: String
+    { id                :: String
+    , tenant_id         :: String
+    , name              :: String
+    , description       :: String
+    , url_slug          :: String
     , tags              :: String
-    , created_at_min    :: String
-    , created_at_max    :: String
-    , updated_at_min    :: String
-    , updated_at_max    :: String
-    , limit             :: String
-    , limit_count       :: Int
-    , offset            :: Int
-    , orderby           :: String
-    , fields            :: String
+    , currency          :: String
+    , advertised_price  :: Int
+    , comparison_price  :: Int
+    , cost_price        :: Int
+    , product_type      :: String
+    , is_published      :: Bool
+    , properties        :: String
     } deriving (Generic)
 
 instance FromJSON Product 
 
-type ProductResult = String
-type ProductAPI = "products" :>
-                (Capture "product_id" Int :>
-                    QueryParam "fields" String :>
+data Field = Field String
+
+instance FromHttpApiData Field where
+    parseUrlPiece x = fmap Field (parseUrlPiece x :: Either Text String)
+
+
+data ProductId = ProductId Int 
+
+instance FromHttpApiData ProductId where
+    parseUrlPiece x = fmap ProductId (parseUrlPiece x :: Either Text Int)
+
+data Sku = Sku String
+
+instance FromHttpApiData Sku where
+    parseUrlPiece x = fmap Sku (parseUrlPiece x :: Either Text String)
+
+data ProductType = ProductType String
+
+instance FromHttpApiData ProductType where
+    parseUrlPiece x = fmap ProductType (parseUrlPiece x :: Either Text String)
+
+data ProductOrdering = ProductOrdering Int 
+
+instance FromHttpApiData ProductOrdering where
+    parseUrlPiece x = fmap ProductOrdering (parseUrlPiece x :: Either Text Int)
```

Comment:
Is there any way to reduce the boilerplate related to `FromHttpApiData`?

/cc @wz1000